### PR TITLE
WIP dbUpdate add transaction support

### PR DESCRIPTION
**Pull request by bofh80**
First attempt at adding transaction support to avoid SQL lnterrupted and deadlocks. (i think)
I haven't tested this as yet, looking for comments first.

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: includes/dbFacile.php [hunk 1]
```diff
@@ -210,8 +210,9 @@ function dbUpdate($data, $table, $where = null, $parameters = [])
     }
 
     try {
-        $result = Eloquent::DB()->update($sql, (array)$data);
-
+        $result = Eloquent::DB()->transaction(function () {
```
Comment (Jellyfrog) [inline includes/dbFacile.php:213]: Missing `use($data, $sql)`

**Comment by Jellyfrog** [comment]
I really doubt we want to use transactions for every single insert

**Comment by mcineran** [comment]
I see this repeated in the logs:
```
[2020-08-31 19:10:39] production.ERROR: SQLSTATE[70100]: <<Unknown error>>: 1317 Query execution was interrupted (SQL: UPDATE `ports` set `ifInOctets`=306459307,`ifInOctets_prev$
#1 /opt/librenms/includes/polling/functions.inc.php(315): include('/opt/librenms/i...')
#2 /opt/librenms/poller.php(142): poll_device(Array, false)
#3 {main}  
[2020-08-31 19:10:39] production.ERROR: SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction (SQL: UPDATE `ports_stati$
#1 /opt/librenms/includes/polling/functions.inc.php(315): include('/opt/librenms/i...')
#2 /opt/librenms/poller.php(142): poll_device(Array, false)
#3 {main}
```
For various updates. from the docs, this seemed like the obvious to do:
https://laravel.com/docs/7.x/database#database-transactions

So is there another way to avoid or at least *help* with the above errors ?

File: includes/dbFacile.php (hunk 1 above)
Comment (mcineran) [reply includes/dbFacile.php:213]: Do I just need to add that line in?
$result = Eloquent::DB()->transaction(function () {
    use($data, $sql)
Or does it need to be in function() or after that call? (apologies my php is not strong)

Comment (bofh80) [reply includes/dbFacile.php]: I believe i've updated this correctly, but now it's working on PHP 7.2 and 7.4.  But failing on PHP 7.3.
Before the change it was the other way around.

Comment (bofh80) [reply includes/dbFacile.php]: Fixed, thanks! that sorted the travis bulds.

**Comment by Jellyfrog** [comment]
I never seen that. I think you should post your issue on the forum instead. 
Also your log is cut off.

**Comment by mcineran** [comment]
Apologies it seems I should have mentioned a couple of things. I'm runing this with a galera backend. 2 hosts, each with a poller, each pointing at their own host. (with other stuff done for redis memcached etc).
So I assume this is the cause, but I need the pollers and hosts to be HA.
When I run with this code i do still see errors:
```
[2020-09-01 22:08:01] production.ERROR: There is already an active transaction (SQL: UPDATE `alerts` set `alerted`=1 WHERE rule_id = 22 && device_id = 62)#0 /opt/librenms/LibreNMS/Alert/RunAlerts.php(483): dbUpdate(Array, 'alerts', 'rule_id = ? && ...', Array)
#1 /opt/librenms/alerts.php(53): LibreNMS\Alert\RunAlerts->runAlerts()
#2 {main}  
```
What I'm unsure of at this point (i've been reading the docs) is if this alert fired once (as per the retries) logged the alert, then succeeded, so no further errors. OR if it logs the above when the retries are exhausted, and should repeat X amount of retries in the logs (the same event repeated X times).

From the docs it seems that this might be logged on each attempt. But if anyone can claify? :)

**Comment by mcineran** [comment]
Yeah, just to confirm if anyone comes across this before, due to my redis backend. Fixed that and this issue went away.

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/mariadb-galera-cluster-general-error-1205-lock-wait-timeout-exceeded/14201/2

**Pull request closed without merging**